Improve auto changelog script

Gudahtt opened the pull request:
The auto changelog script was creating empty or invalid entries in a number of different cases, such as when the body of a commit spanned multiple lines. This has been fixed, and the following additional improvements have been made:

- Error handling (it will now crash upon encountering an error)
- Commits without a PR number in the subject are listed without the PR
prefix
- Invalid shellcheck warnings ignored
- Only the first line of the commit body is shown
- Carriage returns are stripped (some commits contain them)

This script should be more reliable for helping to manually update the changelog. It's still not sufficiently robust to use as part of an automated process - I don't think that's feasible without maintaining stricter control over commit messages conventions and/or merge strategies.

Gudahtt committed:
Improve auto changelog script

The auto changelog script was creating empty or invalid entries in a
number of different cases, such as when the body of a commit spanned
multiple lines. This has been fixed, and the following additional
improvements have been made:

- Error handling (it will now crash upon encountering an error)
- Commits without a PR number in the subject are listed without the PR
prefix
- Invalid shellcheck warnings ignored
- Only the first line of the commit body is shown
- Carriage returns are stripped (some commits contain them)

This script should be more reliable for helping to manually update the
changelog. It's still not sufficiently robust to use as part of an
automated process - I don't think that's feasible without maintaining
stricter control over commit messages conventions and/or merge
strategies.

Gudahtt commented:
Also, I have only tested this on Linux. Somebody should probably test this on macOS before we merge this.

whymarrh commented:
I will test this out on macOS

whymarrh reviewed (changes requested):
Can we run this through ShellCheck one more time and, in particular, address [SC2196](https://github.com/koalaman/shellcheck/wiki/SC2196)?

Gudahtt commented:
Whoops, I must have made that change last and forgot to re-run it. Thanks, that has been updated.

I also added another SC2001 ignore comment for the `pr` variable assignment on line 23. Trying to do that using parameter expansion seems overly complex.

Gudahtt commented on development/auto-changelog.sh:
```diff
@@ -1,26 +1,50 @@
+        prefix=''
     fi
+
+    body="$(git show -s --format="%b" "$commit" | head -n 1 | tr -d '\r')"
```
Oops - this is incorrect.

In the case where the subject lacks a PR ID and there is a commit body, this will use the first line of the body as the description instead of the subject.

Gudahtt committed:
Update script to handle normal commits properly

Commits that were neither Merge commits nor the result of a `Squash &
Merge` were being incorrectly handled - the subject was discarded, and
the first line of the body used as the description instead.

The commit subject is now used as the description for these commits
instead. A fallback for merge commits without a body has been added as
well.

whymarrh commented on development/auto-changelog.sh:
```diff
@@ -1,26 +1,63 @@
-#! /bin/bash
-# update tags
+#!/bin/env bash
```
```suggestion
#!/usr/bin/env bash
```

macOS doesn't have `/bin/env` but some Linux-es do? 🤷🏽‍♂️

Gudahtt replied:
Oops, yes that was a mistake.

whymarrh commented on development/auto-changelog.sh:
```diff
@@ -1,26 +1,63 @@
+
+        first_line_of_body="$(git show -s --format="%b" "$commit" | head -n 1 | tr -d '\r')"
+
+        if grep -E -q '\(#[[:digit:]]+\)' || [[ -z "$first_line_of_body" ]]
```
Is this invocation of `grep` missing an input?

Gudahtt replied:
Oops, yes it is, there should be a `$subject` there.

whymarrh reviewed (changes requested):
A few quick comments

whymarrh commented on development/auto-changelog.sh:
```diff
@@ -1,26 +1,63 @@
+    if grep -E -q '\(#[[:digit:]]+\)|#[[:digit:]]+\sfrom' <<< "$subject"
+    then
+        # shellcheck disable=SC2001
+        pr="$(printf "%s" "$subject" | sed 's/^.*#\([[:digit:]]\+\).*$/\1/')"
```
This doesn't quite work on macOS. 

`GNU bash, version 4.4.19(1)-release (x86_64-pc-linux-gnu)`:

```
# subject='Merge pull request #6745 from MetaMask/fix-readme'
# pr="$(printf "%s" "$subject" | sed 's/^.*#\([[:digit:]]\+\).*$/\1/')"
# printf '%s\n' "$pr"
6745
```

`GNU bash, version 5.0.0(1)-release (x86_64-apple-darwin17.7.0)`:

```
$ subject='Merge pull request #6745 from MetaMask/fix-readme'
$ pr="$(printf "%s" "$subject" | sed 's/^.*#\([[:digit:]]\+\).*$/\1/')"
$ printf '%s\n' "$pr"
Merge pull request #6745 from MetaMask/fix-readme
```

We might be able to use AWK here instead of sed? The following works on macOS:

```suggestion
        pr="$(awk '{sub("#", "", $4); print $4}' <<< $subject)"
```

whymarrh replied:
Although below we're using `#$pr` twice and `$pr` once so we could maybe switch its value to having the `#` prefix (just `awk '{print $4}' <<< $subject`) and then stripping it (`"${pr###}"`) when we don't need it.

Gudahtt replied:
This line is meant to grab the PR id from both "Squash & Merge" and "Merge" cases though. For "Squash & Merge" the _last_ field has the id rather than the fourth.

Gudahtt replied:
I think it would be cleaner to deal with those two cases separately. Then we could use `awk` as you suggested, which seems far simpler than whatever's going on with `sed`.

I was surprised to see `sed` fail like that - maybe it's the escape character? The default on Linux is to escape special characters, but there's a flag to invert that. Maybe the inverse is default on macOS.

whymarrh commented:
I just realized that I have Bash 4 on my macOS machine, before I approve I'll test this out with the stock Bash 3 that macOS currently ships with as the default.

whymarrh commented:
The default macOS Bash:

```bash
$ /bin/bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin17)
Copyright (C) 2007 Free Software Foundation, Inc.
```

whymarrh commented on development/auto-changelog.sh:
```diff
@@ -1,26 +1,62 @@
+
+most_recent_tag="$(git describe --tags "$(git rev-list --tags --max-count=1)")"
+
+readarray -t commits <<< "$(git rev-list "${most_recent_tag}"..HEAD)"
```
`readarray` doesn't work in Bash 3. We might be able to use this kind of setup for the loop:

```diff
@@ -10,9 +10,7 @@ git fetch --tags
 
 most_recent_tag="$(git describe --tags "$(git rev-list --tags --max-count=1)")"
 
-readarray -t commits <<< "$(git rev-list "${most_recent_tag}"..HEAD)"
-
-for commit in "${commits[@]}"
+git rev-list "${most_recent_tag}"..HEAD | while read commit
 do
     subject="$(git show -s --format="%s" "$commit")"
 
```

Gudahtt replied:
😞 `readarray` is great! macOS users using default Bash have it rough.

The `while read` loop is a good suggestion though; that's better than what we had before. We don't need the array.

whymarrh commented on development/auto-changelog.sh:
```diff
@@ -1,26 +1,62 @@
+    # Squash & Merge: the commit subject is parsed as `<description> (#<PR ID>)`
+    if grep -E -q '\(#[[:digit:]]+\)' <<< "$subject"
+    then
+        pr="$(awk '{print $NF}' <<< "$subject")"
```
Looks like this still has the parens, we might be able to use:

```suggestion
        pr="$(awk '{print $NF}' <<< "$subject" | tr -d '()')"
```

This works for me.

Gudahtt replied:
Huh. Not sure how I missed this. That looks like a good solution 👍

whymarrh reviewed (changes requested):
Two comments but otherwise looks good. Tested on Bash 3 & 4 on macOS and Bash 4 Ubuntu.

Gudahtt committed:
Trim 'Squash & Merge' PR id parens

Co-Authored-By: Whymarrh Whitby <whymarrh.whitby@gmail.com>

Gudahtt committed:
Use `while read` instead of `readarray`

`readarray` doesn't work in Bash 3. We didn't really need the commits in
an array anyway, we just needed to loop through them, so `while read`
is a better solution regardless.

whymarrh reviewed (approved):
LGTM, thanks @Gudahtt! This is awesome work! 🎉 

(And sorry about the slow review.)
